File: terminal/src/com/jediterm/terminal/model/JediTermDebouncerImpl.java
package com.jediterm.terminal.model;

import com.jediterm.typeahead.Debouncer;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class JediTermDebouncerImpl implements Debouncer {
  private static final ScheduledExecutorService myScheduler = Executors.newScheduledThreadPool(1);
  private final Object myLock = new Object();

  private final Runnable myRunnable;
  private final long myDelay;

  private TimerTask myTimerTask = null;

  public JediTermDebouncerImpl(Runnable runnable, long delay) {
    myRunnable = runnable;
    myDelay = delay;
  }

  @Override
  public void call() {
    synchronized (myLock) {
      if (myTimerTask != null) {
        myTimerTask.cancel();
      }

      myTimerTask = new TimerTask();
      myScheduler.schedule(myTimerTask, myDelay, TimeUnit.NANOSECONDS);
    }
  }

  @Override
  public void terminateCall() {
    synchronized (myLock) {
      if (myTimerTask != null) {
        myTimerTask.cancel();
        myTimerTask = null;
      }
    }
  }

  private class TimerTask implements Runnable {
    private final long myDueTime;
    private volatile boolean myIsActive = true;

    public TimerTask() {
      myDueTime = System.nanoTime() + myDelay;
    }

    public void cancel() {
      myIsActive = false;
    }

    public void run() {
      if (!myIsActive) {
        return;
      }

      long remaining = myDueTime - System.nanoTime();
      if (remaining > 0) { // Re-schedule task
        myScheduler.schedule(this, remaining, TimeUnit.NANOSECONDS);
      } else { // Mark as terminated and invoke callback
        myIsActive = false;
        myRunnable.run();
      }
    }
  }
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/model/JediTermDebouncerImpl.kt
package com.jediterm.terminal.model

import com.jediterm.core.typeahead.Debouncer
import com.jediterm.terminal.TerminalExecutorServiceManager
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import kotlin.concurrent.Volatile

class JediTermDebouncerImpl(
    private val myRunnable: Runnable, private val myDelay: Long,
    executorServiceManager: TerminalExecutorServiceManager
) : Debouncer {
    private val myLock = Any()

    private val myScheduler: ScheduledExecutorService

    private var myTimerTask: TimerTask? = null

    init {
        myScheduler = executorServiceManager.singleThreadScheduledExecutor
    }

    override fun call() {
        synchronized(myLock) {
            myTimerTask?.cancel()
            myTimerTask = TimerTask()
            myScheduler.schedule(myTimerTask, myDelay, TimeUnit.NANOSECONDS)
        }
    }

    override fun terminateCall() {
        synchronized(myLock) {
            myTimerTask?.cancel()
            myTimerTask = null
        }
    }

    private inner class TimerTask : Runnable {
        private val myDueTime: Long

        @Volatile
        private var myIsActive = true

        init {
            myDueTime = System.nanoTime() + myDelay
        }

        fun cancel() {
            myIsActive = false
        }

        override fun run() {
            if (!myIsActive) {
                return
            }

            val remaining = myDueTime - System.nanoTime()
            if (remaining > 0) { // Re-schedule task
                myScheduler.schedule(this, remaining, TimeUnit.NANOSECONDS)
            } else { // Mark as terminated and invoke callback
                myIsActive = false
                myRunnable.run()
            }
        }
    }
}
